src/hardware/link-mso19/protocol.h

```diff
@@ -158,12 +159,70 @@ SR_PRIV void stop_acquisition(const struct sr_dev_inst *sdi);
 #define REG_PT_MASK(x)		(x + 4)
 #define REG_PT_SPIMODE		8
 
+/* bits - REG_STATUS */
+#define BITS_STATUS_ACTION(x)		(x & STATUS_MASK)
+enum {
+	STATUS_MASK =			0x7,
+	STATUS_NOT_RUNNING =		1,
+	STATUS_TRIGGER_PRE_FILL =	3,
+	STATUS_TRIGGER_WAIT =		4,
+	STATUS_TRIGGER_POST_FILL =	5,
+	STATUS_DATA_READY =		6,
+	BIT_STATUS_ARMED =		1 << 4,
+	BIT_STATUS_OK =			1 << 5,
+};
+
+/* bits - REG_TRIG */
+enum {
+	TRIG_THRESH_MSB_MASK =	3 << 0,
+
+	TRIG_EDGE_RISING =	0 << 2,
+	TRIG_EDGE_FALLING =	1 << 2,
+	TRIG_EDGE_T_F =		0 << 2,
+	TRIG_EDGE_F_T =		1 << 2,
+	TRIG_EDGE_MASK =	1 << 2,
+
+	TRIG_OUT_TRIGGER =	0 << 3,
+	TRIG_OUT_PG =		1 << 3,
+	TRIG_OUT_NOISE =	3 << 3,
+	TRIG_OUT_MASK =		3 << 3,
+
+	TRIG_SRC_DSO =		0 << 5,
+	TRIG_SRC_DSO_PULSE_GE =	1 << 5,
+	TRIG_SRC_DSO_PULSE_LT =	2 << 5,
+	TRIG_SRC_SPI =		4 << 5,
+	TRIG_SRC_I2C =		5 << 5,
+	TRIG_SRC_LA =		7 << 5,
+	TRIG_SRC_MASK =		7 << 5,
+};
+#define TRIG_UPDATE_MASK(reg, val, mask) reg = (((reg) & ~(mask)) | ((val) & (mask)))
+#define TRIG_UPDATE_THRESH_MSB(reg, val) TRIG_UPDATE_MASK((reg), (val) >> 8, TRIG_THRESH_MSB_MASK)
+#define TRIG_UPDATE_EDGE(reg, val)	TRIG_UPDATE_MASK((reg), (val), TRIG_EDGE_MASK)
+#define TRIG_UPDATE_OUT(reg, val)	TRIG_UPDATE_MASK((reg), (val), TRIG_OUT_MASK)
+#define TRIG_UPDATE_SRC(reg, val)	TRIG_UPDATE_MASK((reg), (val), TRIG_SRC_MASK)
```

Comment:
I'd like to suggest to remove most of the macros from `protocol.h`: They are not type save, hide functionality from the developer and are hard to maintain.
The `TRIG_UPDATE_MASK` macro is in my opinion a no go, because it additionally contains an assignment to one of its parameters.